dlt/destinations/path_utils.py

```diff
@@ -1,45 +1,207 @@
-# this can probably go some other place, but it is shared by destinations, so for now it is here
-from typing import List, Sequence, Tuple
+import re
+from typing import Any, Dict, List, Optional, Sequence, Set, Tuple
 
 import pendulum
-import re
 
-from dlt.destinations.exceptions import InvalidFilesystemLayout, CantExtractTablePrefix
+from dlt.cli import echo as fmt
+from dlt.common.storages.load_package import ParsedLoadJobFileName
+from dlt.destinations.exceptions import (
+    CantExtractTablePrefix,
+    InvalidFilesystemLayout,
+    InvalidPlaceholderCallback,
+)
+from dlt.destinations.impl.filesystem.typing import TCurrentDateTime
+
+
+# For formatting options please see
+# https://github.com/sdispater/pendulum/blob/master/docs/docs/string_formatting.md
+DATETIME_PLACEHOLDERS = {
+    # Years
+    "YYYY",  # 2024, 2025, 2026
+    "YY",  # 00, 01, 02 ... 12, 13
+    "Y",  # 2024, 2025, 2026
+    # Months
+    "MMMM",  # January, February, March
+    "MMM",  # Jan, Feb, Mar
+    "MM",  # 01-12
+    "M",  # 1-12
+    # Days
+    "DD",  # 01-31
+    "D",  # 1-31
+    # Hours
+    "HH",  # 00-23
+    "H",  # 0-23
+    # Minutes
+    "mm",  # 00-59
+    "m",  # 0-59
+    # Days of week
+    "dddd",  # Monday, Tuesday, Wednesday
+    "ddd",  # Mon, Tue, Wed
+    "dd",  # Mo, Tu, We
+    "d",  # 0-6
+    # Quarters of the year
+    "Q",  # 1, 2, 3, 4
+}
+
+SUPPORTED_PLACEHOLDERS = DATETIME_PLACEHOLDERS.union(
+    {
+        "schema_name",
+        "table_name",
+        "load_id",
+        "file_id",
+        "ext",
+        "curr_date",
+        "timestamp",
+    }
+)
 
-# TODO: ensure layout only has supported placeholders
-SUPPORTED_PLACEHOLDERS = {"schema_name", "table_name", "load_id", "file_id", "ext", "curr_date"}
 
 SUPPORTED_TABLE_NAME_PREFIX_PLACEHOLDERS = ("schema_name",)
 
 
-def check_layout(layout: str) -> List[str]:
+def get_placeholders(layout: str) -> List[str]:
+    return re.findall(r"\{(.*?)\}", layout)
+
+
+def prepare_datetime_params(
+    current_datetime: Optional[pendulum.DateTime] = None,
+    load_package_timestamp: Optional[str] = None,
+) -> Dict[str, str]:
+    params: Dict[str, str] = {}
+    now = pendulum.now()
+    current_timestamp: pendulum.DateTime = None
+    if load_package_timestamp:
```

Comment:
1. just one timestamp
2. `load_package_timestamp` just for a timestamp from a state